CliFx/Suggestions/Design.md

```diff
@@ -0,0 +1,272 @@
+# then: the user is presented with a series of options to select from:
+PS C:\> cli.exe book
+book        book add        book list       book remove
+```
+
+### Powershell (ubuntu):
+* tab to view first auto-completion
+* tab twice to see possible auto-completions
+```ps
+# when: user presses tab for auto-complete
+PS C:\> cli.exe <tab>
+# then: auto-complete fills in the very first match
+PS C:\> cli.exe book
+# when: user continues to press tab...
+PS C:\> cli.exe book<tab>
+# then: the user is presented with all options
+book        book add        book list       book remove
+PS C:\> cli.exe book
+```
+
+### Bash (ubuntu):
+* tab to view first auto-completion
+* tab twice to see possible auto-completions
+```bash
+# when: user presses tab for auto-complete
+mauricel@DESKTOP: ~/$ cli <tab>
+# then: auto-complete fills in the very first match
+mauricel@DESKTOP: ~/$ cli book
+# when: user continues to press tab...
+mauricel@DESKTOP: ~/$ cli book<tab>
+# then: the user is presented with all options
+book        book add        book list       book remove     ...plus any files present
+mauricel@DESKTOP: ~/$ cli
+```
+
+# Architecture
+
+With CliFx, the auto-complete feature is comprised of the following components
+
+* terminal hook. This is typically an addition to a user's profile. The hook
+  * identifies the command to trigger upon 
+  * extracts cursor position and command line arguments 
+  * calls CliFx to determine the suggestions to provide suggestions
+* CliFx implements the `[suggest]` directive, which is called as follows:
+
+```PS
+# for CliFx.Demo book add title_of_book
+
+CliFx.Demo.exe [suggest] CliFx.Demo book add title_of_book
+# completion suggestions provided, one per line, in standard output
+```
+* The command line, and the cursor position can also be supplied by environment variable for more accurate results.
+
+```PS
+# for CliFx.Demo book add title_of_book
+
+CliFx.Demo.exe [suggest] CliFx.Demo --envar <name_of_variable> --cursor <index_of_cursor>
+# completion suggestions provided, one per line, in standard output
+```
+
+## Configuration
+
+The `[suggest]` directive (or suggest mode) is configured with CliApplicationBuilder as follows. For now, it is disabled by default.
+
+```cs
+new CliApplicationBuilder()
+                // other configuration here.
+                .AllowSuggestMode(true)
+                .Build()
+                .RunAsync();
+```
+
+## Installation
+Installation consists of updating user profile to invoke CliFx suggest mode. Installation can be invoked by user on a per-user basis as follows:
+
+```PS
+cli [suggest] --install
+# powershell refresh
+& $profile
+
+# bash refresh
+source ~/.bashrc
+```
+The following environments are supported
+
+* Windows Powershell, detected by the presence of "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe"
+* Ubuntu Powershell, detected by the presence of "/usr/bin/pwsh"
+* Ubuntu Bash, detected by the presence of "/usr/bin/bash"
```

Comment:
If I have Ubuntu and both Powershell and Bash installed, what happens if I run `cli [suggest] --install`? Will it install on both or just Powershell?

Also, what if I'm using an entirely different terminal? I guess the command will execute successfully, but then autocomplete won't work, which may confuse the user.